### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by zinagithub**

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
+
+class Player {
+	constructor(name, symbol) {
+		this.name = name,
+		this.symbol = symbol
+	}
+}
+
+const Board = ( () => {
+	const boardArray = [null, null, null, null, null, null, null, null, null];
```
Comment (rvvergara) [inline script.js]: There's a cleaner way to do this:

```javascript
const boardArray = new Array(9).fill(null);
```

File: script.js [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+					[3,4,5],
+					[6,7,8],
+					[0,3,6],
+					[1,4,7],
+					[2,5,8],
+					[0,4,8],
+					[6,4,2]
+					];
+    
+	function render() {
```
Comment (rvvergara) [inline script.js]: Let's not mix DOM related logic with our `Board` factory. Our board factory's methods should only include:

1. A function that updates the `boardArray` depending on player's choice
2. A function that checks whether a player's current set of moves already is within any array in  `winCombinations`
3. A function that checks whether or not `boardArray` is full (checking if none of the elements are null).

File: script.js [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+		const boardDiv = document.getElementById('board');
+		const cells = document.getElementsByClassName("sqr");
+		if (cells.length === 0){
+			for (let i=0; i<boardArray.length; i++) {
```
Comment (rvvergara) [inline script.js]: When working with `JavaScript` arrays, it's generally better to use `forEach`, `for of`,`map` or other default methods than the usual `for` loop.

File: script.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+			for (let i=0; i< cells.length; i++){
+				cells[i].innerText = "";
+				cells[i].style.removeProperty('background-color');
+			}
+		}
+		
+		
+	}
+
+	function checkWinner(symb){
```
Comment (rvvergara) [inline script.js]: A simpler way to do this is for each player object to have an array consisting of all the choices the player already has made. And then we can use `every` and `some` array methods to check whether any `winCombinations` inner array elements would be contained within that player's set of moves.

File: script.js [hunk 5]
```diff
@@ -0,0 +1,231 @@
+
+class Player {
```
Comment (rvvergara) [inline script.js:2]: Our player lacks some functionality and feature. For one to make our `Board` methods simpler our player needs to have a storage of all its made moves. Another is the player object should have a method that returns the index a user chooses.

**Review by rvvergara — CHANGES_REQUESTED** [review]
Hi @zinagithub and @YellowPipe,

## WOW WOW WOW!! Very nicely done user interface!!!

### For the 1-player mode can we make it more exciting? :D Haha it's so easy to beat the computer

## Codebase

We need to do some re-engineering of our `Player` and `Board` factories. Please see my annotations.

File: script.js (hunk 3 above)
Comment (zinagithub) [reply script.js]: Hi Ryanv,
I want to tell you that my partner and I did almost the suggestions that you made in your last review only the suggestion about the player wasn't clear  for us, you ask to store the player moves in the player object, in this case i will have an arrays for each player it means that i will have a lot of empty spots, can you be more explicit?
thanks.

File: script.js (hunk 5 above)
Comment (zinagithub) [reply script.js:2]: Hi Ryan,
I want to tell you that my partner and I did almost the suggestions that you made in your last review only the suggestion about the player wasn't clear  for us, you ask to store the player moves in the player object, in this case i will have an arrays for each player it means that i will have a lot of empty spots, can you be more explicit?
thanks.

Comment (rvvergara) [reply script.js:2]: I see you converted your Player into a class. That's not what I meant.

Ok consider this suggested factory for Player:

```javascript
const Player = (name, symbol) => {
  const moves = []; // this is an array where we store the indices made by the player
                     // during the course of the game
 
// the following function is what happens whenever the player makes a move
 const turn = (index, board) => {
    moves.push(index);
    board.markSquare(index, symbol);
  }

 return {
   name,
  symbol,
  moves,
  turn
 }
}
```

File: script.js [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+	constructor(name, symbol) {
+		this.name = name,
+		this.symbol = symbol	
+	}
+}
+
+const Board = ( () => {
```
Comment (rvvergara) [inline script.js:24]: Considering my suggestion above for Player. We can simplify our board factory as such:

```javascript
const Board = () => {
  const boardArray = new Array(9).fill(null);

  const winCombinations = // same as what you already have

  const markSquare = (index, symbol) => {
    if(boardArray[index] === null){
          boardArray[index] = symbol;
    }
  };
  
  const checkWinner = (player) => {
    return winCombinations.some(combo => combo.every(element =>   player.moves.includes(element)));
  }

const checkDraw = () => {
    return boardArray.every(element => element !== null);
  }
}
```

**Review by rvvergara — CHANGES_REQUESTED** [review]
Hi @zinagithub and @YellowPipe,

I don't wanna give you exact code on what I wanna suggest but I just needed to do it (see script.js) so that you will fully grasp the recommendation. You can customize based on how your game flow goes. But suffice to say what we needed to happen was a simpler board factory by providing simpler methods for it. Also the player factory has to do a job other than have a name and symbol. 

Please consider the suggested changes and let me know what you think.

File: script.js (hunk 6 above)
Comment (zinagithub) [reply script.js:24]: Thanks raynv, now it's more explicit.

File: script.js [hunk 7]
```diff
@@ -0,0 +1,219 @@
+
+class Player {
+	constructor(name, symbol) {
+		this.name = name,
+		this.symbol = symbol,
+		this.moves = []
+	}
+
+	checkWin (board) {
```
Comment (rvvergara) [inline script.js:9]: @YellowPipe and @zinagithub, 

After sharing this last review:
```javascript
const checkWinner = (player) => {
    return winCombinations.some(combo => combo.every(element =>   player.moves.includes(element)));
  }
```
we still built a very complex `checkWin` method here. I already demonstrated the use of `every` and `some` yet we still used a nested `for` loop here.

**Review by rvvergara — APPROVED** [review]
Hi @zinagithub and @YellowPipe,

Generally I recommend the thorough practice of OOP and SOLID concepts. It's indeed very challenging to build a truly SOLID codebase at the start but practice makes excellent (not perfect, there's no such a thing existing in the flesh). 

You guys have built a working app and it's a great start. Let's just keep refining our skills to become better and excellent software engineers. Let us get used to building apps that not only work but also are easily extendable and easy to collaborate in. 

Having said this, to not tie you up any longer on this project I will now be giving you a:

# :+1:

**Pull request merged**